prep/go/tutorial/chapter4/xkcd.go

```diff
@@ -0,0 +1,465 @@
+            termMatch := ""
+            termMatchDistance := 100
+            for indexTerm, _ := range index {
+                distance := lev(term, indexTerm)
+                if (distance < termMatchDistance) {
+                    termMatchDistance = distance
+                    termMatch = indexTerm
+                }
+            }
+            // Once term is matched add comic ids
+            if termMatch != "" {
+                for comicId, _ := range index[strings.ToLower(termMatch)] {
+                    idsMapToWeight[comicId]++
+                }
+            }
+        }
+    }
+    for id, weight := range idsMapToWeight {
+        weightedIds = append(weightedIds, WeightedId{id, weight})
+    }
+    // TODO Stack overflowed this but it is pretty easy to understand
+    // Worth reading more about sort module though
+    sort.Slice(weightedIds, func(i, j int) bool {
+        return weightedIds[i].Weight > weightedIds[j].Weight
+    })
+    idsToReturn := min(3, len(weightedIds))
+    for i := 0; i < idsToReturn; i++ {
+        ids = append(ids, weightedIds[i].Id)
+    }
+    return ids[:idsToReturn]
+}
+
+// Creates a simple case-insensitive inverted index of alphabetic terms
+func loadAndBuildIndex(comics *[]XKCDComic) ComicIndex {
+    indexData, err := os.ReadFile(COMIC_INDEX_STORAGE)
+    if err != nil {
+        if os.IsNotExist(err) {
+            indexData = []byte("{}")
+        } else {
+            log.Fatal(err)
+            os.Exit(1)
+        }
+    } 
+    index := newComicIndex()
+    err = json.Unmarshal(indexData, &index)
+    if err != nil {
+        log.Fatal(err)
+        os.Exit(1)
+    }
+
+    for _, comic := range *comics {
+        if !comic.Indexed {
+            // index title and transcript
+            textToIndex := strings.Join([]string{comic.SafeTitle, comic.Title, comic.Transcript}, " ")
+            scanner := bufio.NewScanner(strings.NewReader(textToIndex))
+            scanner.Split(scanLetters)
+            for scanner.Scan() {
+                term := strings.ToLower(scanner.Text())
+                // Skip words like "a" and "to"
+                if len(term) < 3 {
+                    continue
+                }
+                addComicIndex(index, term, comic.Id)
+            }
+
+            // TODO flush comics back to disk with Index = True so it
+            // doesn't get indexed every run
+            // Trying to do it as everything is written now introduces a bug
+            // that isn't worth fixing until the performance is desired (profiling exercise? :))
+            comic.Indexed = true 
+        }
+    }
+    flushIndexToDisk(&index)
+    return index
+}
+
+// TODO ChatGPT'd this. Need to understand writing custom scanners outside this exercise
+func scanLetters(data []byte, atEOF bool) (advance int, token []byte, err error) {
+    start := 0
+    for ; start < len(data); start++ {
+        if unicode.IsLetter(rune(data[start])) {
```

Comment:
I think this will only work if everything is ASCII; `rune(data[index])` isn't a valid way to get Unicode characters out of a `[]byte`.